nav2_map_server/include/nav2_map_server/map_saver.hpp

```diff
@@ -16,56 +17,77 @@
 #define NAV2_MAP_SERVER__MAP_SAVER_HPP_
 
 #include <string>
-#include "map_mode.hpp"
-#include "nav_msgs/srv/get_map.hpp"
+#include <memory>
+
 #include "rclcpp/rclcpp.hpp"
+#include "nav2_util/lifecycle_node.hpp"
+#include "nav2_msgs/srv/save_map.hpp"
+
+#include "mapio.hpp"
 
 namespace nav2_map_server
 {
 
 /**
  * @class nav2_map_server::MapSaver
- * @brief A class that writes map to a file from occpancy grid which is
- * subscribed from "/map" topic.
+ * @brief A class that provides map saving methods and services
  */
-class MapSaver : public rclcpp::Node
+class MapSaver : public nav2_util::LifecycleNode
 {
 public:
   /**
-   * @brief Constructor for the MapSaver
-   * @param options NodeOptions for the MapSaver
+   * @brief Constructor for the nav2_map_server::MapSaver
    */
-  explicit MapSaver(const rclcpp::NodeOptions & options);
+  MapSaver();
 
   /**
-   * @brief A Map callback function calls try_write_map_to_file method to
-   * write map data to a file.
-   * @param map Occupancy Grid message data
+   * @brief Destructor for the nav2_map_server::MapServer
    */
-  void mapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr map);
+  ~MapSaver();
 
   /**
-   * @brief Returns the promise as shared future
-   * @return a shared future copy of save_next_map_promise
+   * @brief Read a message from incoming map topic and save map to a file
+   * @param map_topic Incoming map topic name
+   * NOTE: map_topic could be updated during function execution.
+   * @param save_parameters Map saving parameters.
+   * NOTE: save_parameters could be updated during function execution.
+   * @return true of false
    */
-  std::shared_future<void> map_saved_future() {return save_next_map_promise.get_future().share();}
+  bool saveMapTopicToFile(std::string & map_topic, SaveParameters & save_parameters);
 
 protected:
-  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::ConstSharedPtr map_sub_;
+  // Lifecycle interfaces
+  nav2_util::CallbackReturn on_configure(const rclcpp_lifecycle::State & state) override;
```

Comment:
Can you add doxygen for each? They can be copy pasted for each with 1 word change but just something.

Reply:
Done starting from https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-e22d88faf6e8ac7a5338c8c6cbc2f7a4R58